File: server/src/playlist/playlist.service.ts
import {Injectable} from '@nestjs/common';
import {CreatePlaylistDto} from './dto/create-playlist.dto';
import {Model, ObjectId} from 'mongoose';
import {Playlist, PlaylistDocument} from './schemas/playlist.schema';
import {FileService, FileType} from '../file/file.service';
import {InjectModel} from '@nestjs/mongoose';
import {Song, SongDocument} from "../song/schemas/song.schema";
import * as mongoose from "mongoose";

@Injectable()
export class PlaylistService {
  constructor(
    @InjectModel(Playlist.name) private playlistModel: Model<PlaylistDocument>,
    @InjectModel(Song.name) private songModel: Model<SongDocument>,
    private fileService: FileService,
  ) {}


  async create(
    dto: CreatePlaylistDto,
    userId,
    userName,
    picture,
  ): Promise<Playlist> {
    const picturePath = this.fileService.createFile(FileType.IMAGE, picture);

    const [playlist] = await Promise.all([
      this.playlistModel.create({
        ...dto,
        user: userId,
        userName: userName,
        picture: picturePath,
        public: false,
      }),
    ]);

    return playlist;
  }

  async getOne(id: ObjectId): Promise<Playlist> {
    const [playlist] = await Promise.all([
      this.playlistModel.findById(id).populate('songs'),
    ]);
    return playlist;
  }

  async addSong(id: ObjectId, songId: ObjectId) {
    const playlist = await this.playlistModel.findById(id);
    const song = await this.songModel.findById(songId);
    playlist.songs.push(song._id);
    await playlist.save();
    return playlist;
  }

  async delete(id: ObjectId) {
    const playlist = await this.playlistModel.findById(id);
    await playlist.remove();
    return playlist;
  }

  async getUser(userId: string) {
    const playlists = await this.playlistModel.find({user: new mongoose.Types.ObjectId(userId)});
    return playlists;
  }

  async makePublic(id: ObjectId) {
    return this.playlistModel.findByIdAndUpdate(id, {public: true});
  }
}

File: backend/src/main/java/com/muz/service/PlaylistService.java
package com.muz.service;

import com.muz.document.PlaylistDocument;
import com.muz.document.SongDocument;
import com.muz.dto.CreatePlaylistDto;
import com.muz.exception.ApiException;
import com.muz.repository.PlaylistRepository;
import com.muz.repository.SongRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class PlaylistService {

    private final PlaylistRepository playlistRepository;
    private final SongRepository songRepository;
    private final FileStorageService fileStorage;

    public PlaylistDocument create(CreatePlaylistDto dto, String userId, String userName, MultipartFile picture) {
        String picturePath = fileStorage.store("image", picture);

        PlaylistDocument playlist = new PlaylistDocument();
        playlist.setName(dto.getName());
        playlist.setUser(userId);
        playlist.setUserName(userName);
        playlist.setPicture(picturePath);
        playlist.setPub(false);
        return playlistRepository.save(playlist);
    }

    public Map<String, Object> getOne(String id) {
        PlaylistDocument playlist = playlistRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Playlist not found"));
        List<SongDocument> songs = songRepository.findAllById(playlist.getSongs());
        return Map.of("playlist", playlist, "songs", songs);
    }

    public PlaylistDocument addSong(String playlistId, String songId) {
        PlaylistDocument playlist = playlistRepository.findById(playlistId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Playlist not found"));
        songRepository.findById(songId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Song not found"));

        if (!playlist.getSongs().contains(songId)) {
            playlist.getSongs().add(songId);
        }
        return playlistRepository.save(playlist);
    }

    public String delete(String id) {
        playlistRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Playlist not found"));
        playlistRepository.deleteById(id);
        return id;
    }

    public List<PlaylistDocument> getByUser(String userId) {
        return playlistRepository.findByUser(userId);
    }

    public PlaylistDocument makePublic(String id) {
        PlaylistDocument playlist = playlistRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Playlist not found"));
        playlist.setPub(true);
        return playlistRepository.save(playlist);
    }
}
